addons/digest/models/digest.py

```diff
@@ -55,20 +56,72 @@ def _compute_available_fields(self):
+        companies = self.company_id
+
+        if any(not digest.company_id for digest in self):
+            # No company: we will use the current company to compute the KPIs
+            companies |= self.env.company
+
+        return (
+            fields.Datetime.to_string(self._context.get('start_datetime')),
+            fields.Datetime.to_string(self._context.get('end_datetime')),
+            companies,
+        )
 
     def _compute_kpi_res_users_connected_value(self):
-        for record in self:
-            start, end, company = record._get_kpi_compute_parameters()
-            user_connected = self.env['res.users'].search_count([('company_id', '=', company.id), ('login_date', '>=', start), ('login_date', '<', end)])
-            record.kpi_res_users_connected_value = user_connected
+        self._compute_company_based_kpi(
+            'res.users',
+            'kpi_res_users_connected_value',
+            date_field='login_date',
+        )
 
     def _compute_kpi_mail_message_total_value(self):
+        start, end, __ = self._get_kpi_compute_parameters()
         discussion_subtype_id = self.env.ref('mail.mt_comment').id
-        for record in self:
-            start, end, company = record._get_kpi_compute_parameters()
-            total_messages = self.env['mail.message'].search_count([('create_date', '>=', start), ('create_date', '<', end), ('subtype_id', '=', discussion_subtype_id), ('message_type', 'in', ['comment', 'email'])])
-            record.kpi_mail_message_total_value = total_messages
+        self.kpi_mail_message_total_value = self.env['mail.message'].search_count([
+            ('create_date', '>=', start),
+            ('create_date', '<', end),
+            ('subtype_id', '=', discussion_subtype_id),
+            ('message_type', 'in', ('comment', 'email')),
+        ])
+
+    def _compute_company_based_kpi(self, model, field_name, date_field='create_date',
+                                   domain=None, sum_field=None):
+        """Generic method that computes the KPI on a given model.
+
+        :param model: Model on which we will count the number of record
+            This model must have a "company_id" field
+        :param field_name: Field name on which we will write the KPI
+        :param date_field: Field used for the date range
+        :param domain: Additional domain for the search
+        :param sum_field: Field to sum to obtain the KPI,
+            if None it will count the number of records
+        """
+        start, end, companies = self._get_kpi_compute_parameters()
+
+        search_domain = [
+            ('company_id', 'in', companies.ids),
+            (date_field, '>=', start),
+            (date_field, '<', end),
+        ]
+
+        if domain:
+            search_domain = expression.AND([search_domain, domain])
+
+        values = self.env[model]._read_group(
+            domain=search_domain,
+            fields=[f'{sum_field}:sum' if sum_field else 'id'],
```

Comment:
Do you even need this part if sum_field is not defined? I thought it did that automatically.